devolutions-gateway/src/api/fwd.rs

```diff
@@ -237,3 +254,288 @@ where
+
+    let host_value =
+        header::HeaderValue::from_str(claims.target_host.host()).map_err(HttpError::bad_request().err())?;
+    request.headers_mut().insert(header::HOST, host_value);
+
+    // 4. Forward the request.
+
+    let response = if matches!(request.uri().scheme_str(), Some("ws" | "wss")) {
+        // 4.a Prepare the WebSocket upgrade.
+
+        // We are discarding the original body.
+        // There is no HTTP body when performing a WebSocket upgrade.
+        let (mut parts, _) = request.into_parts();
+
+        let client_ws = WebSocketUpgrade::from_request_parts(&mut parts, &state).await.map_err(
+            HttpError::bad_request()
+                .with_msg("failed to initiate the websocket upgrade")
+                .err(),
+        )?;
+
+        // 4.b Open a WebSocket connection to the target.
+
+        let request = axum::http::Request::from_parts(parts, ());
+        let request_uri = request.uri().clone();
+
+        let tls_connector = if dangerous_tls {
+            Some(DANGEROUS_TLS_CONNECTOR.clone())
+        } else {
+            None
+        };
+
+        let (server_ws, server_ws_response) = connect_async_tls_with_config(request, None, false, tls_connector)
+            .await
+            .map_err(
+                HttpError::bad_gateway()
+                    .with_msg("WebSocket connection to target server")
+                    .err(),
+            )?;
+
+        let server_stream = tokio_tungstenite_websocket_compat(server_ws);
+
+        debug!(?server_ws_response, %dangerous_tls, "Connected to target server");
+
+        // 4.c Start WebSocket forwarding.
+
+        let span = tracing::Span::current();
+        let conf = state.conf_handle.get_conf();
+        let sessions = state.sessions;
+        let subscriber_tx = state.subscriber_tx;
+
+        client_ws.on_upgrade(move |client_ws| {
+            let client_stream = crate::ws::websocket_compat(client_ws);
+            let client_addr = source_addr;
+
+            async move {
+                info!(target = %request_uri, "WebSocket-WebSocket forwarding");
+
+                let info = SessionInfo::builder()
+                    .association_id(claims.jet_aid)
+                    .application_protocol(claims.jet_ap)
+                    .details(ConnectionModeDetails::Fwd {
+                        destination_host: claims.target_host,
+                    })
+                    .time_to_live(claims.jet_ttl)
+                    .recording_policy(claims.jet_rec)
+                    .build();
+
+                let result = Proxy::builder()
+                    .conf(conf)
+                    .session_info(info)
+                    .address_a(client_addr)
+                    .transport_a(client_stream)
+                    .address_b("8.8.8.8:8888".parse().expect("valid hardcoded value")) // NOTE: We don’t really use this address for anything else other than pcap recording, so it’s fine to use a placeholder for now.
```

Comment:
[nit][style] This line with the comment looks pretty long, could we move the address to constant and add a comment above?